From 6fdad8196d0a2648fd5c354096342e582c3322c2 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Tue, 2 Jan 2024 11:35:59 +0100
Subject: [PATCH 0289/1695] drivers: stm32_rif: add
 stm32_rif_access_violation_action()

This function should be used by peripherals capable on raising
access violation interrupts (SERC, IAC). The behavior of the platform
on such event is platform-specific. Therefore, its definition must be
done at platform level.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 8cbeaf81465e0c1212fd241e3eeb52e5de6e06ca drivers: stm32_rif: add stm32_rif_access_violation_action()]
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Change-Id: Ie941d22f5e149b05b38688bf8ac13cbf6d1796d5
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/333123
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c | 15 +++++++++++++++
 core/drivers/firewall/stm32_iac.c  |  9 ++-------
 core/include/drivers/stm32_iac.h   |  6 ------
 core/include/drivers/stm32_rif.h   | 13 +++++++++++++
 4 files changed, 30 insertions(+), 13 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index e6acc7267..ba048fddf 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -6,6 +6,9 @@
 #include <config.h>
 #include <console.h>
 #include <drivers/gic.h>
+#include <drivers/stm32_rif.h>
+#include <drivers/stm32_risab.h>
+#include <drivers/stm32_risaf.h>
 #include <drivers/stm32_uart.h>
 #include <initcall.h>
 #include <kernel/boot.h>
@@ -149,3 +152,15 @@ void boot_secondary_init_intc(void)
 {
 	gic_cpu_init();
 }
+
+#ifdef CFG_STM32_RIF
+void stm32_rif_access_violation_action(void)
+{
+#ifdef CFG_STM32_RISAF
+	stm32_risaf_dump_erroneous_data();
+#endif
+#ifdef CFG_STM32_RISAB
+	stm32_risab_dump_erroneous_data();
+#endif
+}
+#endif /* CFG_STM32_RIF */
diff --git a/core/drivers/firewall/stm32_iac.c b/core/drivers/firewall/stm32_iac.c
index ee6889788..2d63ab6c4 100644
--- a/core/drivers/firewall/stm32_iac.c
+++ b/core/drivers/firewall/stm32_iac.c
@@ -3,6 +3,7 @@
  * Copyright (c) 2022-2024, STMicroelectronics
  */
 #include <drivers/stm32_iac.h>
+#include <drivers/stm32_rif.h>
 #include <dt-bindings/soc/stm32mp25-rifsc.h>
 #include <io.h>
 #include <kernel/boot.h>
@@ -123,12 +124,6 @@ static TEE_Result stm32_iac_parse_fdt(void)
 #define IAC_ILAC_ID(reg_val, offset)	(IAC_FIRST_ILAC_IN_REG(reg_val) + \
 					 IAC_EXCEPT_LSB_BIT(offset))
 
-__weak __noreturn void access_violation_action(void)
-{
-	DMSG("Ooops...");
-	panic();
-}
-
 static const uint32_t iac_fbd_periph[] = {
 	STM32MP25_RIFSC_PWR_ID, STM32MP25_RIFSC_GPIOA_ID,
 	STM32MP25_RIFSC_GPIOB_ID, STM32MP25_RIFSC_GPIOC_ID,
@@ -200,7 +195,7 @@ static enum itr_return stm32_iac_itr(struct itr_handler *h __unused)
 	}
 
 	if (do_panic)
-		access_violation_action();
+		stm32_rif_access_violation_action();
 
 	return ITRR_HANDLED;
 }
diff --git a/core/include/drivers/stm32_iac.h b/core/include/drivers/stm32_iac.h
index 5c6341fcd..b926516d7 100644
--- a/core/include/drivers/stm32_iac.h
+++ b/core/include/drivers/stm32_iac.h
@@ -20,11 +20,5 @@ struct stm32_iac_platdata {
 	int irq;
 };
 
-/*
- * Platform can define an access violation action to trap cpu/reset core or system
- * after an eventual debug trace.
- */
-void access_violation_action(void);
-
 int stm32_iac_get_platdata(struct stm32_iac_platdata *pdata);
 #endif /* __DRIVERS_STM32_IAC_H__ */
diff --git a/core/include/drivers/stm32_rif.h b/core/include/drivers/stm32_rif.h
index a72c38254..99ad6881e 100644
--- a/core/include/drivers/stm32_rif.h
+++ b/core/include/drivers/stm32_rif.h
@@ -8,6 +8,7 @@
 
 #include <drivers/stm32mp_dt_bindings.h>
 #include <io.h>
+#include <kernel/panic.h>
 #include <mm/core_memprot.h>
 #include <util.h>
 
@@ -178,4 +179,16 @@ static inline TEE_Result
 }
 #endif
 
+/*
+ * The action to take upon an access violation depends on the platform.
+ * Therefore, it should be defined at platform level
+ */
+#ifdef CFG_STM32_RIF
+void stm32_rif_access_violation_action(void);
+#else /* CFG_STM32_RIF */
+static inline void stm32_rif_access_violation_action(void)
+{
+}
+#endif /* CFG_STM32_RIF */
+
 #endif /* __DRIVERS_STM32_RIF_H */
-- 
2.47.2

